### ISSUE TITLE
[20.03 regression]: autologin does not work for "root"

### ISSUE THREAD

**Issue by volth** [NixOS/nixpkgs#81651]
**Describe the bug**

In `19.09` there was `services.xserver.displayManager.auto.enable` option which did "root" (or any other user) autologin, which was very useful for VM automation

In `20.09` it was removed proposing to replace it with
```
displayManager.lightdm.enable           = true;
displayManager.lightdm.autoLogin.enable = true;
displayManager.lightdm.autoLogin.user   = "root";
```

But it rejects to do auto login for "root" user!

Adding an unpriveleged user and setting its name to `displayManager.lightdm.autoLogin.user` does the trick, but it would require a refactoring much bigger than adding 2 extra lines (it would require review all the automation scripts adding "sudo" here and there)

**To Reproduce**
1.
make a VM with
```
services.xserver = rec {
        windowManager.icewm.enable = true;
        displayManager.defaultSession = "none+icewm";
        displayManager.lightdm.enable           = true;
        displayManager.lightdm.autoLogin.enable = true;
        displayManager.lightdm.autoLogin.user   = "root";
}
```

**Expected behavior**
![](https://imgur.com/rm6xCmX.png)

**Screenshots**
![](https://imgur.com/xaJogjx.png)

**Comment by worldofpeace** [comment]
@volth Please read the release notes, it outlines this in detail and what you can do to get exact functionality without the module.

**Comment by worldofpeace** [comment]
https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2003.xml#L497

**Comment by worldofpeace** [comment]
I don't believe any display manager is going to ever allow this default in the pam rules, or that we should include any option in nixos to mutate this. Closing.

**Comment by worldofpeace** [comment]
You could also just import the module from our testing infra https://github.com/NixOS/nixpkgs/blob/master/nixos/tests/common/auto.nix. If it's just for integration testing with the driver it would make sense.

**Comment by volth** [comment]
>  we should include any option in nixos to mutate this

But... the option is already there: `services.xserver.displayManager.lightdm.autoLogin.user`.
Why shouldn't it fail when set to "root" ?

**Comment by volth** [comment]
My idea is not to add extra option, but to remove `test-support.displayManager.auto.enable` and `test-support.displayManager.auto.user` which are there only to work around the broken case of 
`services.xserver.displayManager.lightdm.autoLogin.user = "root"`

Why not fix `services.xserver.displayManager.lightdm.autoLogin.user` and deprecate that `test-support.*` hack?

Wouldn't it be more consistent?

**Comment by worldofpeace** [comment]
> > we should include any option in nixos to mutate this
> 
> But... the option is already there: `services.xserver.displayManager.lightdm.autoLogin.user`.
> Why shouldn't it work when set to "root" ?

Because all the pam rule's default forbid this. Perhaps all the autoLogin options should specify only normalUsers can be autologined. And with wayland this is completely impossible, and it's why we've changed the default autologin user in the installation media not to be root https://github.com/NixOS/nixpkgs/commit/397c7d26fcb001ce5e1e3c53a3366524c4f91bf9

**Comment by worldofpeace** [comment]
> `playManager.auto.enable` and `test-support.displayManager.auto.user` which are there only to work around the broken case of
> `services.xserver.displayManager.lightdm.autoLogin.user = "root"`
> 
> Why not fix `services.xserver.displayManager.lightdm.autoLogin.user` and deprecate that `test-support.*` hack (together with a kilobyte of prose it requires to explain) ?
> 
> Wouldn't it be more consistent?

By rule of least privillege is why the PAM rules haven't been like this for a long long time now.
The "prose" I've used to describe how to achieve this is actually duplication of information you can easily find on the internet. See this wiki with info from **Fedora 10** https://fedoraproject.org/wiki/Enabling_Root_User_For_GNOME_Display_Manager. The usecase could best be supported by moving the information into the nixos.wiki.

**Comment by worldofpeace** [comment]
(and this usecase has no guarantee of continuing to work because wayland)

**Comment by volth** [comment]
Fedora is different.

We have declarative configurations here.

I declare a configuration and it does not work.

Intentionally.
 
And it is faster to fix than to to find and read all the comments and wiki which tries explain that it is forbidden because it is forbidden in Fedora.

ohh.

**Comment by worldofpeace** [comment]
> We have declarative configurations here.
> I declare a configuration and it does not work.

What's in the manual is basically a translation of this to a declarative nixos configuration.
There's nothing stopping anyone from overriding the PAM module configuration, similar to just edit a file in place on Fedora. 

```
security.pam.services.lightdm-autologin.text = lib.mkForce ''
    auth     requisite pam_nologin.so
    auth     required  pam_succeed_if.so quiet
    auth     required  pam_permit.so

    account  include   lightdm

    password include   lightdm

    session  include   lightdm
'';

**Comment by conferno** [comment]
similar to just edit a file in place on Fedora.
> 
> ```
> security.pam.services.lightdm-autologin.text = lib.mkForce ''
>     auth     requisite pam_nologin.so
>     auth     required  pam_succeed_if.so quiet
>     auth     required  pam_permit.so
> 
>     account  include   lightdm
> 
>     password include   lightdm
> 
>     session  include   lightdm
> '';
> ```

I see it as a very weird solution.
It is already duplicated 2 times across `<nixpkgs>` (and probably already went out of sync, as there are more differences than `uid >= 1000`) and you offer to copy-paste it more

**Comment by worldofpeace** [comment]
I don't understand what you mean by "it is already duplicated 2 times across nixpkgs", it's exactly used in both lightdm and gdm.
Most display-managers offer an autologin functionality, so they have an autologin pam module configuration. The state of the duplication is already known https://github.com/NixOS/nixpkgs/issues/17044, I'm not offering to copy-paste more. This configuration above is not copy-pasted. It's a copy paste **and** without the user uid clause.

**Comment by worldofpeace** [comment]
The substance of the PR was not to support it at all, it wasn't. The auto display manager module existed because of its need in internal testing infra https://github.com/NixOS/nixpkgs/commit/c95612a5a2d4bd93011c042066c6c3deacfd436d. If someone wanted to do root autologin in any other display manager it wouldn't have worked, and they would have had to override the pam configuration. That's exactly what the auto module did in the first place.

**Comment by conferno** [comment]
> I don't understand what you mean

1. original
https://github.com/NixOS/nixpkgs/blob/8dc5ff7dcfd1c58c32004ffae25e6d31ed83d86c/nixos/modules/services/x11/display-managers/lightdm.nix#L302-L314

2. override
https://github.com/NixOS/nixpkgs/blob/c95612a5a2d4bd93011c042066c6c3deacfd436d/nixos/tests/common/auto.nix#L52-L64

There are more differences than checking uid.
Which of them are related to root autologin and which to internal testing?
Should one dig into that just to log in as root?

**Comment by worldofpeace** [comment]
Hmm, maybe I didn't look into the differences closely since I just moved the module out :rofl: 
Perhaps it's different because it was based off the module before I rewrote it.
But I'm not really going to reiterate autologin as root being supported by default, or made easy.
For the very visible community consensus from display-manager developers, it's a bad idea. That's why I mentioned this was a thing even in Fedora 10 (and by extension that meant from GDM developers because that's their reference distro).

**Comment by volth** [comment]
> root being supported by default, or made easy.

I did not expect it from NixOS - intentionally making something difficult to configure (even if dangerous things - it is not a commercial consumer OS after all, where such position could be justified by cutting costs on L1 support).

**Comment by worldofpeace** [comment]
> I did not expect it from NixOS - intentionally making something difficult to configure

? This module was created in 2010 https://github.com/NixOS/nixos/commit/73668f41dafd3570987fef58808f125a1534c8c7, and around that time DM devs were intentionally concerned about the root user. The fact that we created a module when it was meant for internal testing was a mistake IMHO, because it wasn't actually supported, and obfuscated the fact this module used lightdm (there's no such thing as an `auto` display manager project).

If a user had no idea the `auto` module existed, if you look at the pam configurations, logging in as root wouldn't work. 

I believe this change is very much a no-op, nothing was made intentionally difficult to do (by us). It was already like this if you overlook the auto display-manager module's existence. Migration was documented for a topic that is ancient in internet time. I'm going to now withdraw from the discussion in this thread because I believe I supported my reasoning up to the questioning enough.

**Comment by volth** [comment]
> It was already like this if you overlook the auto display-manager module's existence.

offtopic, but: `auto` display-manager was cool and not confusing at all: pure the idea of "I do not care of DM brands, whether it is slim now or `lightdm` in the next NixOS. When a machine starts, just run my IceWM with the configured user".

**Comment by nh2** [comment]
I came here after some hours of debugging why `displayManager.lightdm.autoLogin.user = "root";` doesn't work.

We used it to declare a VM image that immediately logs into a desktop to make a workflow easily reproducible. That seems to no longer work.

Questions:

* If it's intentionally not working, could the `autoLogin.user` setting describe that some users are forbidden in there and will silently not work?
* Can anybody explain what I need to do to make our VM work again as intended? Set `security.pam.services.lightdm-autologin.text` to one of the values (which one?) from https://github.com/NixOS/nixpkgs/issues/81651#issuecomment-594210008?

**Comment by nh2** [comment]
Still relevant to me.
